airbyte-connector-builder-server/connector_builder/impl/default_api.py

```diff
@@ -2,21 +2,344 @@
+        """
+        Using the provided manifest and config, invokes a sync for the specified stream and returns groups of Airbyte messages
+        that are produced during the read operation
+        :param stream_read_request_body: Input parameters to trigger the read operation for a stream
+        :return: Airbyte record messages produced by the sync grouped by slice and page
+        """
+        return self.temp_read_stream(stream_read_request_body)
+
+    def temp_read_stream(self, stream_read_request_body: StreamReadRequestBody = Body(None, description="")) -> StreamRead:
+        try:
+            adapter = self._create_low_code_adapter(stream_read_request_body)
+        except ValidationError as error:
+            return KnownExceptionInfo(
+                message=f"Invalid connector manifest with error: {error.message}", exception_class_name=ValidationError.__name__
+            )
+
+        single_slice = StreamReadSlices(pages=[])
+        log_messages = []
+
+        first_page = True
+        current_records: List[object] = []
+        current_page_request: Optional[HttpRequest] = None
+        current_page_response: Optional[HttpResponse] = None
+        try:
+            # Message groups are partitioned according to when request log messages are received. Subsequent response log messages
+            # and record messages belong to the prior request log message and when we encounter another request, append the latest
+            # message group
+            for message in adapter.read_stream(stream_read_request_body.stream, stream_read_request_body.config):
+                if first_page and message.type == Type.LOG and message.log.message.startswith("request:"):
+                    first_page = False
+                    request = self._create_request_from_log_message(message.log)
+                    current_page_request = request
+                elif message.type == Type.LOG and message.log.message.startswith("request:"):
+                    if not current_page_request or not current_page_response:
+                        self.logger.warning("Every message grouping should have at least one request and response")
+                    single_slice.pages.append(
+                        StreamReadPages(request=current_page_request, response=current_page_response, records=current_records)
+                    )
+                    current_page_request = self._create_request_from_log_message(message.log)
+                    current_records = []
+                elif message.type == Type.LOG and message.log.message.startswith("response:"):
+                    current_page_response = self._create_response_from_log_message(message.log)
+                elif message.type == Type.LOG:
+                    log_messages.append({"message": message.log.message})
+                elif message.type == Type.RECORD:
+                    current_records.append(message.record.data)
+            else:
```

Comment:
the indentation looks off here?

Reply:
actually that was intended. The `else` at the end of the for loop just means we'll execute this statement when the loop finishes iterating. Which we want so that we append the final page grouping to the slice.